### DISCUSSION TITLE
Cannot customize rich text link base fields

### DISCUSSION THREAD

**Discussion by evelynhathaway** [payloadcms/payload#2167] [Roadmap]
From what I can tell, there isn't a way to customize the fields that the link modal that displays in rich text fields. The custom fields and the native fields are isolated from each other.

https://github.com/payloadcms/payload/blob/baf5b10d23306e3c6bd5ad43c6e0b9b48cf7c334/src/admin/components/forms/field-types/RichText/elements/link/LinkDrawer/baseFields.ts#L16

https://github.com/payloadcms/payload/blob/25c5b784e119c638d0c9d25fee4309df02244861/src/admin/components/forms/field-types/RichText/elements/link/Button/index.tsx#L71-L89

This means users can't:

- Create custom validation for the URL field (e.g. disallowing links to certain domains)
- Change the collections that the are referenced in the internal link option https://github.com/payloadcms/payload/discussions/2095 (e.g. removing collections that do not have URLs like `Users`)
- Minimize the amount of data that is returned by the internal link option (per answer from the Payload team in https://github.com/payloadcms/payload/discussions/1960)
- Remove the option for either internal links or custom URLs
- Remove the open in new tab checkbox

**Comment by jmikrut** [comment]
Hey @evelynhathaway — this is a great call. It should also be pretty easy to accomplish.

I think what we need to do is expose a property on the richText field as follows:

```
field.admin.link.overrideDefaultFields = (defaultFields) => newFields
```

Basically, it'd be a function that would accept default fields and allow Payload devs to return them modified to suit their purpose.

I don't think that this would be a good way to filter down the data returned still though... That might need to be done with an additional function - say something like:

``` 
field.reduceRelationshipData = ({ doc, slug }) => doc
```

This function could be shared and reused with any relationship populated by the rich text field. It could also automatically type the `doc` arg via the collection `slug` argument, like we do within our local API methods. It would also type the returned `doc` as a `Partial`.

We'd run this function each time we populate an `upload` or `relationship` field within rich text.

What do you think? This is a great feature. Not necessarily a bug or an issue, so I will move this to discussions, but I think this should be built ASAP.

Speaking of moving this to a discussion, I will add this to the roadmap as Priority 2 (to be released within next quarter or sooner). But in the meantime we would happily accept a PR for this. If you're up for it, I am happy to help provide more info on how this would need to be accomplished.

👍

**Reply by foopis23** [reply]
Has anyone started work on this? I would like to take an attempt if not.

**Reply by jmikrut** [reply]
Hey @foopis23 we have not! And help on this one would be AMAZING. The team and I are happy to help you in-depth on Discord if you end up giving this a shot. 👍

**Reply by ChrisGV04** [reply]
Hi @jmikrut I could give it a try. I would just like to know what kind of behavior would be ideal for these cases. I believe that it would be nice if you can choose to keep the default fields, add extra fields below the defaults or override them entirely with your own fields for maximum control. So I propose the following:

1. There could be a `disableDefaults: true` property inside the `admin.link` to completely disable the default fields and override them with your own inside `admin.link.fields`. This way we can set our own hooks, validations and all that stuff for our fields.
2. If you don't set `disableDefaults: true`, the fields inside `admin.link.fields` will be added under the defaults just as it's currently working.

Edit: This method also makes everything backwards compatible since the current behavior would still be the default unless you use the `disableDefaults: true` property.

Example:

```ts
{
  name: "customLinks",
  type: "richText",
  admin: {
    link: {
      disableDefaults: true, // Remove the default fields
      fields: [
        // ... These fields would override the default ones
      ]
    }
  }
},

{
  name: "linksWithDefaults",
  type: "richText",
  admin: {
    link: {
      fields: [
        // ... These fields would be added below the defaults. Just like the current behaviour
      ]
    }
  }
}
```

Personally, I believe this could be easier for the developer experience than writing a function to override the fields and solve the relationship conflict you mentioned. However, it might be more complicated to implement, but I'm not completely sure to be honest.

If you have any suggestions or believe your proposed solution would be better, please let me know. I'll wait until we agree on a behavior before I try to implement anything.

**Reply by jmikrut** [reply]
Hey there - your notes all make sense. But I think it can be even simpler. What if we just used the existing key `admin.link.fields` and allowed it to accept an array of fields to be used below the default fields, OR a function, that accepts the base fields, and expects you to return an array of all fields you’d like to use? 

I do understand that it might seem easier to specify your own base fields rather than extend the default fields but I also think that using the default fields from the function’s argument would be optional, and you could simply return whatever you want in any case with my proposed solution.

What do you think?

**Reply by ChrisGV04** [reply]
@jmikrut That makes sense. I think I like that approach better. Then if you pass a function you would get the default fields as an argument (which maybe could be typed) and it should return an array of fields, correct?

I imagine that to override a field it would be something like this:

```ts
{
  type: 'richText',
  admin: {
    link: {
      fields: (defaultFields) => {
        // Override the label of the link type field
        defaultFields.linkType.label = 'Type of link to use';

        return [
          ...defaultFields,
          // Add a custom field
          { name: 'showIcon', type: 'checkbox', label: 'Show an icon', defaultValue: false },
        ];
      },
    },
  },
},
```

Did I get it right?

**Reply by foopis23** [reply]
@ChrisGV04 are you going to work on this? Just making sure because I don't want duplicate work on this.

**Reply by ChrisGV04** [reply]
@foopis23 Hey! I'm not currently working on this. So if you want to give it a shot, go ahead 🚀

**Reply by mvarendorff2** [reply]
Since I couldn't find any PRs on this, I have taken the liberty to create one at #2559 :)

**Comment by ChrisGV04** [comment]
Yeah! Maybe we should also be able to disable the default fields entirely (except for the "Text to display"). I have a custom reusable field for the links in my project but when I add it to the Rich Text field's `admin.link.fields` I have duplicated data and the default fields don't fit my needs.

<img width="1180" alt="payload-richtext-link-fields" src="https://user-images.githubusercontent.com/47041342/221089858-9fd67c83-8992-49e1-8433-99f44ef813d6.png">

It would be nice to remove them and use my own fields there.

**Comment by DanRibbens** [comment]
Collection `admin.enableRichTextLink` is now available in Payload v1.7.5. This makes it even easier if all you need to to customize is the omission of certain collections from linking.

**Comment by evelynhathaway** [comment]
Between the following MRs, I think everything from this discussion is complete.

- #2559
- #2756
- #2560

As a note from my implementation, if you replace the default fields (e.g. to make a consistent API with a custom Link `Block`) and still want the rich text popup to work, you will have to write a React component with a side effect that updates the hidden default fields when and of your new ones are changed. In my case, it was way easier to make my Link `Block` match the default rich text field `name`s and adding overrides for all non-`name` properties.

![popup in the rich text editor that says "Linked to Post"](https://github.com/payloadcms/payload/assets/38994283/df56beac-242d-4a15-aea8-2a726b73a8dc)

**Reply by WouterK12** [reply]
Hi, great progress, but I think this discussion is incomplete. (Or this is possibly a feature for another discussion.)
There is still a need for reducing the amount of data returned by relationships.

E.g. like suggested above by @jmikrut:
> I don't think that this would be a good way to filter down the data returned still though... That might need to be done with an additional function - say something like:
> 
> ```
> field.reduceRelationshipData = ({ doc, slug }) => doc
> ```

I can't find this feature on the roadmap. Is it there?

Edit: I think this is related to this discussion: https://github.com/payloadcms/payload/discussions/4896
